### PR TITLE
Initial P1Events page scroll depth tacking

### PR DESCRIPTION
**Pull request by B77Mills**
New P1 Event  tracking components help track page scroll depth.  The intent is to fire p1 events based on how far the user scrolls down the page.  This will use a querySelecort to determine the end of the target page and calculate scroll depth accordingly. 

The example inclusions for the website look something like this at moment  with default events firing at25, 50, 75 & 90%.

### NativeX story:
Here is a [link](http://www-p1-sandbox.dev.parameter1.com:51269/sponsored/parameter1/another-story/62b37e698a17bc0001015d90?preview=true)

**Template:**
```marko.js
<marko-web-p1-events-track-native-story-scroll-depth story=story  selector=".document-container" />
```

**EventsObj:**

```js
{
    "slug": "p1",
    "realm": "5ed294c6c13a4626008b4568",
    "env": "development",
    "host": "www-p1-sandbox.dev.parameter1.com",
    "vis": "FmBWTh5gwTurR0GYdT9It",
    "idt": "omeda.hcl.customer*5234E7727901C9J~encrypted",
    "events": [
        {
            "act": "Scroll",
            "cat": "Content",
            "lab": "25 percent",
            "ent": {
                "id": "native-x.example.story*6309207d5a33e600013c5b83",
                "name": "The ABCs of VIP: How the VIP Experience is Evolving at Sports Venues Across the Country",
                "props": {
                    "type": "story",
                    "published": 1659330000000
                },
                "refs": {
                    "advertiser": {
                        "id": "native-x.example.advertiser*5fcfb9422205c1026a1ebf3e",
                        "name": "Parameter1",
                        "refs": {}
                    }
                }
            }
        }
    ]
},
{
    "slug": "p1",
    "realm": "5ed294c6c13a4626008b4568",
    "env": "development",
    "host": "www-p1-sandbox.dev.parameter1.com",
    "vis": "FmBWTh5gwTurR0GYdT9It",
    "idt": "omeda.hcl.customer*5234E7727901C9J~encrypted",
    "events": [
        {
            "act": "Scroll",
            "cat": "Content",
            "lab": "90 percent",
            "ent": {
                "id": "native-x.example.story*6309207d5a33e600013c5b83",
                "name": "The ABCs of VIP: How the VIP Experience is Evolving at Sports Venues Across the Country",
                "props": {
                    "type": "story",
                    "published": 1659330000000
                },
                "refs": {
                    "advertiser": {
                        "id": "native-x.example.advertiser*5fcfb9422205c1026a1ebf3e",
                        "name": "Parameter1",
                        "refs": {}
                    }
                }
            }
        }
    ]
}
```

### Content Page:
This is a good example of [content pages](http://www-p1-sandbox.dev.parameter1.com:51269/home/article/22288696/how-to-manage-and-reduce-risk-at-large-events)

**Template:**
```marko.js
<marko-web-p1-events-track-content-scroll-depth content=content selector=".document-container .page .row .content-page-body" />
```

**EventsObj:**

```js
{
    "slug": "p1",
    "realm": "5ed294c6c13a4626008b4568",
    "env": "development",
    "host": "www-p1-sandbox.dev.parameter1.com",
    "vis": "FmBWTh5gwTurR0GYdT9It",
    "idt": "omeda.hcl.customer*5234E7727901C9J~encrypted",
    "events": [
        {
            "act": "Scroll",
            "cat": "Content",
            "lab": "25 percent",
            "ent": {
                "id": "base.p1-sandbox.content-article*22288696",
                "name": "How to Manage and Reduce Risk at Large Events",
                "props": {
                    "type": "article",
                    "published": 1655837711000
                },
                "refs": {
                    "primarySection": {
                        "id": "base.p1-sandbox.website-section*64017",
                        "name": "Home",
                        "props": {
                            "alias": "home"
                        },
                        "refs": {}
                    },
                    "company": null,
                    "authors": [],
                    "createdBy": {
                        "id": "base.p1-sandbox.user*5eea498723013d2b008b4571",
                        "name": "jworden",
                        "props": {
                            "firstName": "Josh",
                            "lastName": "Worden"
                        },
                        "refs": {}
                    }
                }
            }
        }
    ]
},
{
    "slug": "p1",
    "realm": "5ed294c6c13a4626008b4568",
    "env": "development",
    "host": "www-p1-sandbox.dev.parameter1.com",
    "vis": "FmBWTh5gwTurR0GYdT9It",
    "idt": "omeda.hcl.customer*5234E7727901C9J~encrypted",
    "events": [
        {
            "act": "Scroll",
            "cat": "Content",
            "lab": "90 percent",
            "ent": {
                "id": "base.p1-sandbox.content-article*22288696",
                "name": "How to Manage and Reduce Risk at Large Events",
                "props": {
                    "type": "article",
                    "published": 1655837711000
                },
                "refs": {
                    "primarySection": {
                        "id": "base.p1-sandbox.website-section*64017",
                        "name": "Home",
                        "props": {
                            "alias": "home"
                        },
                        "refs": {}
                    },
                    "company": null,
                    "authors": [],
                    "createdBy": {
                        "id": "base.p1-sandbox.user*5eea498723013d2b008b4571",
                        "name": "jworden",
                        "props": {
                            "firstName": "Josh",
                            "lastName": "Worden"
                        },
                        "refs": {}
                    }
                }
            }
        }
    ]
}
```

### Website section:

**Template:**
```marko.js
<marko-web-p1-events-track-website-section-scroll-depth section=section selector=".document-container" />
```
**EventObj:**
```js
{
    "slug": "p1",
    "realm": "5ed294c6c13a4626008b4568",
    "env": "development",
    "host": "www-p1-sandbox.dev.parameter1.com",
    "vis": "FmBWTh5gwTurR0GYdT9It",
    "idt": "omeda.hcl.customer*5234E7727901C9J~encrypted",
    "events": [
        {
            "act": "Scroll",
            "cat": "Website Section",
            "lab": "25 percent",
            "ent": {
                "id": "base.p1-sandbox.website-section*83028",
                "refs": {}
            }
        }
    ]
}{
    "slug": "p1",
    "realm": "5ed294c6c13a4626008b4568",
    "env": "development",
    "host": "www-p1-sandbox.dev.parameter1.com",
    "vis": "FmBWTh5gwTurR0GYdT9It",
    "idt": "omeda.hcl.customer*5234E7727901C9J~encrypted",
    "events": [
        {
            "act": "Scroll",
            "cat": "Website Section",
            "lab": "25 percent",
            "ent": {
                "id": "base.p1-sandbox.website-section*83028",
                "refs": {}
            }
        }
    ]
},
{
    "slug": "p1",
    "realm": "5ed294c6c13a4626008b4568",
    "env": "development",
    "host": "www-p1-sandbox.dev.parameter1.com",
    "vis": "FmBWTh5gwTurR0GYdT9It",
    "idt": "omeda.hcl.customer*5234E7727901C9J~encrypted",
    "events": [
        {
            "act": "Scroll",
            "cat": "Website Section",
            "lab": "90 percent",
            "ent": {
                "id": "base.p1-sandbox.website-section*83028",
                "refs": {}
            }
        }
    ]
}
```

### Static & non-standard page calls
**Template:**
```marko.js 
import ns from "@parameter1/base-cms-marko-web-p1-events/utils/create-namespace";
<marko-web-p1-events-track-static-scroll-depth
  entity={
    id: "home",
      ns: ns("website-section"),
        name: "home",
          alias: "home",
  }
  category="Static"
selector = ".document-container"
  />
```
**EventsObj:**
```js
{
  slug: "p1",
  realm: "5ed294c6c13a4626008b4568",
  env: "development",
  host: "www-p1-sandbox.dev.parameter1.com",
  vis: "FmBWTh5gwTurR0GYdT9It",
  idt: "omeda.hcl.customer*5234E7727901C9J~encrypted",
  events: [
    {
      act: "Scroll",
      cat: "Static",
      lab: "25 percent",
      ent: {
        id: "base.p1-sandbox.website-section*home",
        name: "home",
        refs: {}
      }
    }
  ]
  },
  {
    slug: "p1",
    realm: "5ed294c6c13a4626008b4568",
    env: "development",
    host: "www-p1-sandbox.dev.parameter1.com",
    vis: "FmBWTh5gwTurR0GYdT9It",
    idt: "omeda.hcl.customer*5234E7727901C9J~encrypted",
    events: [
      {
        act: "Scroll",
        cat: "Static",
        lab: "90 percent",
        ent: {
          id: "base.p1-sandbox.website-section*home",
          name: "home",
          refs: { }
        }
      }
    ]
  }
];

```

### CODE REVIEW

File: packages/marko-web-p1-events/components/marko.json [hunk 1]
```diff
@@ -37,6 +37,24 @@
     "@action": "string",
     "@category": "string"
   },
+  "<marko-web-p1-events-track-page-view-depth>": {
```
Comment (brandonbk) [inline packages/marko-web-p1-events/components/marko.json]: Rename to `marko-web-p1-events-track-scroll-depth`

File: packages/marko-web-p1-events/components/track-page-view-depth.marko [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import defaultValue from "@parameter1/base-cms-marko-core/utils/default-value";
+import { getAsObject } from "@parameter1/base-cms-object-path";
+import eventEntity from "../utils/base-entity";
+
+$ const obj = getAsObject(input, "obj");
+$ const { selector } = input;
+
+<marko-web-browser-component
```
Comment (brandonbk) [inline packages/marko-web-p1-events/components/track-page-view-depth.marko]: Rename to `P1EventsTrackScrollDepth` and rename file

File: services/example-website/server/templates/content.marko [hunk 3]
```diff
@@ -23,6 +23,7 @@ $ const { recaptcha } = out.global;
         <div class="col-md-8">
           <div class="content-page-body">
             <theme-page-contents|{ blockName }| attrs={ "data-gallery-id": id }>
+            <marko-web-p1-events-track-page-view-depth obj={id: content.id, type: content.type } />
```
Comment (brandonbk) [inline services/example-website/server/templates/content.marko]: Rename to `marko-web-p1-events-track-scroll-depth`

File: services/example-website/server/templates/index.marko [hunk 4]
```diff
@@ -4,6 +4,7 @@
   description="This is a test page."
 >
   <@page>
+    <marko-web-p1-events-track-page-view-depth obj={ id: 'home', type: 'section' } selector=".document-container" />
```
Comment (brandonbk) [inline services/example-website/server/templates/index.marko]: Rename to `marko-web-p1-events-track-scroll-depth`

File: services/example-website/server/templates/section.marko [hunk 5]
```diff
@@ -3,7 +3,9 @@ $ const { recaptcha } = out.global;
 
 <marko-web-website-section-page-layout id=id alias=alias name=name>
   <@page>
+
     <marko-web-resolve-page|{ data: section, resolved }| node=pageNode>
+    <marko-web-p1-events-track-page-view-depth obj={ id: section.id, type: 'Section' } selector=".document-container" />
```
Comment (brandonbk) [inline services/example-website/server/templates/section.marko]: Rename to `marko-web-p1-events-track-scroll-depth`

File: packages/marko-web-p1-events/components/marko.json [hunk 6]
```diff
@@ -37,6 +37,24 @@
...
+    "template": "./track-page-view-depth.marko",
```
Comment (brandonbk) [inline packages/marko-web-p1-events/components/marko.json]: Update reference when file is renamed

File: packages/marko-web-p1-events/browser/track-page-view-depth.vue [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+      type: String,
+      default: '.document-container .page .row .content-page-body',
+    },
+    fullViewDepth: {
+      type: Number,
+      default: 1,
+    },
+    targetViewDepths: {
+      type: Array,
+      default: () => ([0.25, 0.50, 0.75]),
```
Comment (brandonbk) [inline packages/marko-web-p1-events/browser/track-page-view-depth.vue]: Add 0.90 to list of default depths

File: packages/marko-web-p1-events/browser/index.js [hunk 8]
```diff
@@ -1,10 +1,12 @@
 const TrackContentBodyLinks = () => import(/* webpackChunkName: "p1-events-track-content-body-links" */ './track-content-body-links.vue');
+const TrackPageViewDepth = () => import(/* webpackChunkName: "p1-events-track-page-view-depth" */ './track-page-view-depth.vue');
 const TrackDownloadSubmission = () => import(/* webpackChunkName: "p1-events-track-download-submission" */ './track-download-submission.vue');
 const TrackInquirySubmission = () => import(/* webpackChunkName: "p1-events-track-inquiry-submission" */ './track-inquiry-submission.vue');
 
 export default (Browser) => {
   const { EventBus } = Browser;
   Browser.register('P1EventsTrackContentBodyLinks', TrackContentBodyLinks);
+  Browser.register('P1EventsTrackPageViewDepth', TrackPageViewDepth);
```
Comment (brandonbk) [inline packages/marko-web-p1-events/browser/index.js]: Rename to `P1EventsTrackScrollDepth`

File: packages/marko-web-p1-events/browser/track-page-view-depth.vue [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+export default {
+  props: {
+    entity: {
+      type: String,
+      required: true,
+    },
+    selector: {
+      type: String,
+      default: '.document-container .page .row .content-page-body',
```
Comment (brandonbk) [inline packages/marko-web-p1-events/browser/track-page-view-depth.vue]: Remove the default and make the selector required

File: packages/marko-web-p1-events/browser/track-page-view-depth.vue [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    },
+    fullViewDepth: {
+      type: Number,
+      default: 1,
+    },
+    targetViewDepths: {
+      type: Array,
+      default: () => ([0.25, 0.50, 0.75]),
+    },
+    action: {
```
Comment (brandonbk) [inline packages/marko-web-p1-events/browser/track-page-view-depth.vue]: Remove this since it is generated

**Review by brandonbk — DISMISSED** [review]

File: packages/marko-web-p1-events/components/track-scroll-depth/content.marko [hunk 11]
```diff
@@ -0,0 +1,47 @@
+import { getAsArray } from "@parameter1/base-cms-object-path";
+import ns from "../../utils/create-namespace";
+
+$ const { site } = out.global;
+$ const {
+  content,
+  action,
+  category = 'Content',
```
Comment (solocommand) [inline packages/marko-web-p1-events/components/track-scroll-depth/content.marko]: Don't allow action or category to be customized in the specific components -- content should always be content category and action should always be scroll.

The category should be required on your generic/static component, but action should still be hardcoded to scroll.

**Review by solocommand — CHANGES_REQUESTED** [review]

**Review by solocommand — APPROVED** [review]

**Comment by B77Mills** [comment]
[Inbody Scroll depth tracking](https://trello.com/c/YFoNcGcU/242-inbody-scroll-depth-tracking)

**Pull request merged**